### ISSUE TITLE
filterbarcodes TypeError: unhashable type: 'list'

### ISSUE THREAD

**Issue by Brawni** [timoast/sinto#11]
Hi Tim,

I have been running successfully sinto barcodes for a batch of BAM files but now  it's throwing an error with another batch of BAM files. I had a look into the bam and they seem ok, and the barcodes list is formatted the same way. Its possible also that these barcodes won't match anything in the bam, but that doesnt throw any errors.
Here is head of one BAM and the barcode list im trying to subset it to:

```

A00445:16:H7YL5DMXX:1:1439:23601:3098	16	2	46895278	255	1S38M487913N59M	*	GTCACTGCAACCTCCACCTTCCAGGTTCAAGCAATTCTCCTGGGAGGCGGAGCTTGCAGTGAGCCGAGATTGCACCACTGCACTCCAGCCTGGGTGAC	FFFFFFFFFFFFFFFFFFFFFF:FF:FFFFFFFFF:F:FF:FFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:92	nM:i:0	RE:A:I	BC:Z:CCTTTGTC	QT:Z:FFFFFFFF	CR:Z:TTCTCAAAGATGTGGC	CY:Z:FFFFFFFFFFFFFFFF	CB:Z:TTCTCAAAGATGTGGC-1	UR:Z:TTGGGGTTGG	UY:Z::FFFFFFFFF	UB:Z:TTGGGGTTGG	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:1
A00445:16:H7YL5DMXX:1:1236:9136:10614	0	2	47109310	255	66M257308N32M	*	ACTTTGGGAGGCTGATGTGGGTAGATCACCTGAACTCAGGAGTTCAACACCAGCCTGGCCAACAAGAAACCCCATCTCTACTAAAAATACAAAAAATT	:,FFFFFFF,FFF:F,FFFFFFFFFFFFFFFF:F,FFFFFFF:FFFF,F,F:FFF,FFFFFFFF,:FFFFF,FFFFFFFFFFFFF:FFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:84	nM:i:5	RE:A:I	BC:Z:AGCACACT	QT:Z:,:FFFFF,	CR:Z:GTCAGGGAGGTGATAT	CY:Z:F::F,FFFFFFFFF,F	CB:Z:GTCACGGAGGTGATAT-1	UR:Z:ATCTGGGAAG	UY:Z:FFFFFF:F,F	UB:Z:ATCTGGGAAG	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:1
A00445:16:H7YL5DMXX:2:1121:8395:11741	16	2	47151916	255	7S41M257273N50M	*	CGCCGGCTTTGTTTTTTTTTTTTTTTTTGTATTTTTAGTAGAGACAGGGTTTCACCATGTTGGCCAGCCTGGTCTTGAACTCCTGACCTCAAGTGATC	:,F:FF,:,,,:FFFF,FFFFF:FFFFFFFFFFFFFFFFFFFF:FFFFFF,FFFFFFFFFFFFFFF,,FFFFFFFFFF:FFFFFFFFFF:FFF::FFF	NH:i:1	HI:i:1	AS:i:83	nM:i:2	RE:A:I	BC:Z:AGCACACT	QT:Z:FFFFFFFF	CR:Z:ACGGGTCAGCGTGTCC	CY:Z:FFFFFFFFFFFFFFFF	CB:Z:ACGGGTCAGCGTGTCC-1	UR:Z:TGGCGTCTGA	UY:Z:FF:F,:FFFF	UB:Z:TGGCGTCTGA	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:2
A00445:16:H7YL5DMXX:1:1369:20292:30185	0	2	47151961	255	51M200703N47M	*	CAATGTGTTAGCCAGGATGGTCTAGATCTCCTGACCTTGTGATCCGCCCGCCCCTGCCTCCCAAAGTGCTGGGATTACAGGTGTGAGCCACCGTGCCC	FF,FFFFFFFFFFFFFFF:F:FFF:FFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFF:FFFFFF:FFFFFFFFFFFFFFFFFF,FFFFFFFFF	NH:i:1	HI:i:1	AS:i:88	nM:i:3	RE:A:I	BC:Z:AGCAAACT	QT:Z:FFFFFFFF	CR:Z:TGCACGCTCGTGGGAA	CY:Z:FF,FFF:FFFF,FFF:	CB:Z:TGGACGCTCGTGGGAA-1	UR:Z:TGTTCCATTT	UY:Z:FF,:FFFFFF	UB:Z:TGTTCCATTT	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:1
A00445:16:H7YL5DMXX:2:2451:8847:2973	16	2	47168859	255	62M191908N36M	*	CCTCTGCCTCCCAGGTTCAAGTGATTCTCCTGACTCAGCCTCTAGAGTCGCTGGGATTACAGGCACACGCCACCATGCCAGGCTAATTTTTATATTTT	::FFFFF:FFFFFF,FFFFF:FFFFFFF::FF,FFFF:FFFFFFFFFF,FFFF:FFFFFFFF:FFFFFFF,FFF,FFFF,FFFFFFFFFF:FF,:FFF	NH:i:1	HI:i:1	AS:i:84	nM:i:5	RE:A:I	BC:Z:TAGGATGA	QT:Z::F,F,FFF	CR:Z:TGGTTCCAGTCCCGGA	CY:Z:F:FFFF,:,:FFFFFF	CB:Z:TGGTTCCAGTACCGGA-1	UR:Z:ACGTAGAACC	UY:Z:F:FFFFF,::	UB:Z:ACGTAGAACC	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:2
A00445:16:H7YL5DMXX:2:2444:12192:33129	16	2	47179880	255	45M184241N53M	*	CACACCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCGTGCCACCACACCCAGCTAATTTTGTATTTTTAGTAGAGACGGGGTTTC	FFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:90	nM:i:2	RE:A:I	BC:Z:GTACGCGG	QT:Z:FFFFFFFF	CR:Z:AGGCCGTAGCAACGGT	CY:Z:FFFFFFFFFFFFFFFF	CB:Z:AGGCCGTAGCAACGGT-1	UR:Z:TCGCTTTACT	UY:Z:FFFFFFF,FF	UB:Z:TCGCTTTACT	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:2
A00445:16:H7YL5DMXX:1:2449:20229:6840	0	2	47212619	255	22M237731N76M	*	AAACAAAACAAAAAAAAAAAACACCGGGCGTGGTGGCTCACACCTGTAATCCCAGCACTTTGGGAGGCCGAGGCAGGCAGATCACAAGGTCAGGAGAT	F,FFFF:FFFFFFFFFF::FF:FFFFFFFF:FFF,FFFFFFF:F:FFFF,FFFFFFF,FFFFFF:FFFFFFFFFFFF,FF,,FFFFFFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:88	nM:i:3	RE:A:I	BC:Z:AGCACACT	QT:Z:FFFFFFFF	CR:Z:TGCCCTAAGCGTCAAG	CY:Z:FFFFFFFFFFFFFFFF	CB:Z:TGCCCTAAGCGTCAAG-1	UR:Z:GTCACAAATT	UY:Z:FFFFFFFFFF	UB:Z:GTCACAAATT	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:1
A00445:16:H7YL5DMXX:1:2324:16776:36276	16	2	47251321	255	1S24M202853N73M	*	GATCCTCCCTCCTCAGCCTCCCAAAGTGGGCGGATCACGAGGTCAAGACATCAAGACCATCCTGACCAACATGGCGAAACCCCGTCTGTACTAAAAAT	FFFFFFFFFFFFFFFFFFFFFFF:FF,FF,FFFFFFFFFFFFFFFFFF:,FFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFF	NH:i:1	HI:i:1	AS:i:77	nM:i:8	RE:A:I	BC:Z:CCTTTGTC	QT:Z:FFFFFF:F	CR:Z:TGCCCATTCAGAGACG	CY:Z:FFFFF,FFFFFFFFFF	CB:Z:TGCCCATTCAGAGACG-1	UR:Z:GAGTGTGCGA	UY:Z:FFFFFFFFFF	UB:Z:GAGTGTGCGA	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:1
A00445:16:H7YL5DMXX:1:2173:5032:7733	0	2	47253083	255	4S70M108784N24M	*	GGAATTCAAGACCAGCCTGGCCATCATGGTGTAACCCCATCTCTACTAAAAATACTAAAAATTAGCTAGGTGTGGTGGTTCATGCCTGTAATCCCAGC	FFF:FFFFFFFFFFFFFFFFFFF,FFFFFFF,FFF::FFFFFFFFFFFFFFFF:F:F,FFFF,FFFFFFFFFFF:FFFF,F:FFFFFFFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:84	nM:i:3	RE:A:I	BC:Z:GTCCGCGG	QT:Z:FF,:FFFF	CR:Z:GCGGGTTTCTATCCCC	CY:Z:FFFFFFFFFFFFFFFF	CB:Z:GCGGGTTTCTATCCCG-1	UR:Z:ACCTTAGGGG	UY:Z:,FFFFFFFFF	UB:Z:ACCTTAGGGG	RG:Z:RPL_CST:MissingLibrary:1:H7YL5DMXX:1
A00445:16:H7YL5DMXX:2:2344:4227:21245	0	2	47261720	255	42M278937N53M3S	*	GGCTCACACATGTAATCCCAGCACTTTGGGAAGCGAAGGCAGGCGGATTGCTTGAGGCCAGGAGTTTGGGACCAGCCTGGGTCACATAGCCAGACCCT	F,,FFFFFFFFFFFFFFFFFFFFFFF,F,FF:FFF:FFFFFFF,FFFFFFFF:FFFFFFFFFFFFFF:FFFFFF,FF,FFFF,FFF:FFFFFFFFFFF	NH:i:1	HI:i:1	AS:i:72	nM:i:9	RE:A:I	BC:Z:AGCACACT	QT:Z::FF,FFFF	CR:Z:GTGAAGGGTTGTCTTT	CY:Z:FF:FFF

``` 
```

And the barcode tab-delimited file:
AAACCTGAGACTACAA-1	Myeloid
AAACCTGCACCTGGTG-1	Myeloid
AACCATGCATCACGAT-1	T.NK.cells
AACGTTGGTGTTGGGA-1	Myeloid
AACTGGTGTTACGGAG-1	T.NK.cells
AAGACCTTCCAGAGGA-1	T.NK.cells
AAGGAGCTCTGATACG-1	T.NK.cells
AAGGTTCAGGTTACCT-1	T.NK.cells
AAGTCTGGTATCAGTC-1	Myeloid
AAGTCTGTCTATGTGG-1	T.NK.cells

```

And here is the error:
Function run_filterbarcodes called with the following arguments:

```
 sinto filterbarcodes -b file.bam \
 -c celltypes.txt \
--barcodetag "CB"

bam	file.bam
cells	celltypes.txt
trim_suffix	False
nproc	1
barcode_regex	None
barcodetag	CB
func	<function run_filterbarcodes at 0x2b5b48a5b6a8>
Traceback (most recent call last):
  File "/broad/hptmp/bgiotti/signac/bin/sinto", line 263, in <module>
    options.func(options)
  File "/broad/hptmp/bgiotti/signac/lib/python3.6/site-packages/sinto/utils.py", line 21, in wrapper
    func(args)
  File "/broad/hptmp/bgiotti/signac/lib/python3.6/site-packages/sinto/cli.py", line 14, in run_filterbarcodes
    cellbarcode=options.barcodetag,
  File "/broad/hptmp/bgiotti/signac/lib/python3.6/site-packages/sinto/filterbarcodes.py", line 92, in filterbarcodes
    unique_classes = list(set(chain.from_iterable(cb.values())))
TypeError: unhashable type: 'list'

```

Thanks a lot for your support!

**Comment by timoast** [comment]
Hi @Brawni, it looks like this error is happening as we're parsing the celltypes.txt file, not even getting to the stage of iterating through the bam. Could you send me the celltypes.txt file to take a look? I think that would be the fastest solution, the file should be fairly small. You can email it to tstuart@nygenome.org

**Comment by timoast** [comment]
This was due to duplicate lines in the celltypes file, which caused some weird things to happen when parsing the file. I've released a fix now (0.5.1, available on pypi) that handles cases where the same cell barcode is present on multiple line and should fix this issue

**Comment by Brawni** [comment]
Hi Tim,

That works great!

Cheers!
